File: python/tests/test_threads.py
import os
import pytest
from ttio import _threads


def test_resolve_threads_precedence(monkeypatch):
    monkeypatch.delenv("TTIO_THREADS", raising=False)
    monkeypatch.setattr(os, "cpu_count", lambda: 32)
    assert _threads.resolve_threads() == 30
    monkeypatch.setattr(os, "cpu_count", lambda: 4)
    assert _threads.resolve_threads() == 2
    # Two cores would otherwise resolve to zero workers.
    monkeypatch.setattr(os, "cpu_count", lambda: 2)
    assert _threads.resolve_threads() == 1
    monkeypatch.setattr(os, "cpu_count", lambda: 1)
    assert _threads.resolve_threads() == 1
    # 0 means "use the default", which is still the one-core answer here.
    monkeypatch.setenv("TTIO_THREADS", "0")
    assert _threads.resolve_threads() == 1
    monkeypatch.setenv("TTIO_THREADS", "6")
    assert _threads.resolve_threads() == 6
    assert _threads.resolve_threads(2) == 2
    assert _threads.resolve_threads(0) == 6
    monkeypatch.setenv("TTIO_THREADS", "junk")
    assert _threads.resolve_threads() == 1


def test_pool_context_stands_down_the_autotune(monkeypatch):
    calls = []
    monkeypatch.setattr(_threads, "_get_autotune", lambda: 3)
    monkeypatch.setattr(_threads, "_set_autotune", lambda n: calls.append(n))
    with _threads.pool_context(1):
        pass
    assert calls == []
    with _threads.pool_context(4):
        with _threads.pool_context(2):
            assert calls == [1]
        assert calls == [1]
    assert calls == [1, 3]


def test_cli_threads_flag_sets_env(monkeypatch, tmp_path):
    from ttio.tools.workbench_cli import main
    monkeypatch.delenv("TTIO_THREADS", raising=False)
    fq = tmp_path / "in.fastq"
    fq.write_text("@r1\nACGT\n+\nIIII\n")
    assert main(["encode", "--input", str(fq), "--format", "fastq",
                 "--output", str(tmp_path / "o.tio"), "--threads", "3"]) == 0
    assert os.environ["TTIO_THREADS"] == "3"


def test_read_ahead_window_env_override(monkeypatch):
    """The window is settable so the memory/latency trade can be
    measured. Java: GenomicRun.readAheadBlocks; ObjC:
    TTIOReadAheadBlocks."""
    from ttio import genomic_run

    monkeypatch.delenv("TTIO_READ_AHEAD_BLOCKS", raising=False)
    assert genomic_run._read_ahead_blocks() == genomic_run._READ_AHEAD_BLOCKS
    monkeypatch.setenv("TTIO_READ_AHEAD_BLOCKS", "9")
    assert genomic_run._read_ahead_blocks() == 9
    # Junk and non-positive values defer to the default rather than
    # leaving a reader with no lookahead at all.
    monkeypatch.setenv("TTIO_READ_AHEAD_BLOCKS", "junk")
    assert genomic_run._read_ahead_blocks() == genomic_run._READ_AHEAD_BLOCKS
    monkeypatch.setenv("TTIO_READ_AHEAD_BLOCKS", "0")
    assert genomic_run._read_ahead_blocks() == genomic_run._READ_AHEAD_BLOCKS


def test_v6_segment_threads_clamp(monkeypatch):
    """The argument is blocks in flight, not the pool size, and the
    cap is the core count. Parity with Threads.resolveV6SegmentThreads
    (Java) and +[TTIOThreads resolveV6SegmentThreads:] (ObjC)."""
    monkeypatch.setattr(os, "cpu_count", lambda: 32)
    assert _threads.resolve_v6_segment_threads(1) == 32     # the machine
    assert _threads.resolve_v6_segment_threads(4) == 8
    assert _threads.resolve_v6_segment_threads(16) == 2
    assert _threads.resolve_v6_segment_threads(64) == 2     # floor
    assert _threads.resolve_v6_segment_threads(0) == 32     # read as one
    # The product stays near the core count wherever it can.
    for blocks in (1, 2, 4, 8, 16):
        assert _threads.resolve_v6_segment_threads(blocks) * blocks <= 32


def test_v6_segment_threads_env_override(monkeypatch):
    """TTIO_V6_SEGMENT_THREADS wins whatever the worker count, and
    reaches outside the clamp on purpose: a sweep has to be able to ask
    for 1 and for more than 8. Parity with the Java and ObjC resolvers."""
    monkeypatch.setattr(os, "cpu_count", lambda: 32)
    monkeypatch.setenv("TTIO_V6_SEGMENT_THREADS", "5")
    assert _threads.resolve_v6_segment_threads(1) == 5
    assert _threads.resolve_v6_segment_threads(64) == 5
    monkeypatch.setenv("TTIO_V6_SEGMENT_THREADS", "1")
    assert _threads.resolve_v6_segment_threads(1) == 1       # below the floor
    monkeypatch.setenv("TTIO_V6_SEGMENT_THREADS", "16")
    assert _threads.resolve_v6_segment_threads(1) == 16      # above the cap
    for bad in ("junk", "0", "-1", ""):
        monkeypatch.setenv("TTIO_V6_SEGMENT_THREADS", bad)
        assert _threads.resolve_v6_segment_threads(16) == 2  # falls through
    monkeypatch.delenv("TTIO_V6_SEGMENT_THREADS")
    assert _threads.resolve_v6_segment_threads(16) == 2


def test_import_threads_caps_the_default_at_what_memory_affords(monkeypatch):
    # A thread costs a gibibyte of the pipeline budget, so the default
    # count is capped at what a quarter of physical memory affords.
    monkeypatch.delenv("TTIO_THREADS", raising=False)
    monkeypatch.delenv("TTIO_IMPORT_THREADS", raising=False)
    monkeypatch.setattr(os, "cpu_count", lambda: 32)
    per_thread = _threads.IMPORT_BLOCK_BYTES * 16

    def phys(gib):
        monkeypatch.setattr(
            os, "sysconf",
            lambda n: (gib << 30) // 4096 if n == "SC_PHYS_PAGES" else 4096)

    phys(32)                       # 8 GiB quarter / 1 GiB a thread
    assert _threads.resolve_import_threads() == 8
    phys(256)                      # affords more than the plain default
    assert _threads.resolve_import_threads() == 30
    phys(1)                        # affords none: the floor is one
    assert _threads.resolve_import_threads() == 1
    assert (32 >> 2) << 30 == 8 * per_thread


def test_import_threads_honours_what_the_caller_asked_for(monkeypatch):
    monkeypatch.setattr(os, "cpu_count", lambda: 32)
    monkeypatch.setattr(
        os, "sysconf",
        lambda n: (32 << 30) // 4096 if n == "SC_PHYS_PAGES" else 4096)
    monkeypatch.delenv("TTIO_IMPORT_THREADS", raising=False)
    # An explicit count is not capped.
    monkeypatch.setenv("TTIO_THREADS", "30")
    assert _threads.resolve_import_threads() == 30
    # The import knob overrides the rule, and TTIO_THREADS with it.
    monkeypatch.setenv("TTIO_IMPORT_THREADS", "3")
    assert _threads.resolve_import_threads() == 3
    monkeypatch.delenv("TTIO_THREADS")
    assert _threads.resolve_import_threads() == 3
    # Junk falls through to the count asked for.
    monkeypatch.setenv("TTIO_IMPORT_THREADS", "junk")
    monkeypatch.setenv("TTIO_THREADS", "12")
    assert _threads.resolve_import_threads() == 12

File: objc/Tests/TestThreads.m
/* TTIOThreads knob, the auto-tune stand-down, and V6's own segment
 * thread count.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Core/TTIOThreads.h"
#include <stdlib.h>
#include "ttio_rans.h"

void testThreads(void);
void testThreads(void)
{
    unsetenv("TTIO_THREADS");
    NSUInteger cores = (NSUInteger)[[NSProcessInfo processInfo] activeProcessorCount];
    NSUInteger want = cores > 3 ? cores - 2 : 1;
    PASS([TTIOThreads resolve:nil] == want, "threads: default is cores minus 2, at least 1");
    setenv("TTIO_THREADS", "6", 1);
    PASS([TTIOThreads resolve:nil] == 6, "threads: TTIO_THREADS wins over the default");
    PASS([TTIOThreads resolve:@2] == 2, "threads: an explicit value wins");
    PASS([TTIOThreads resolve:@0] == 6, "threads: explicit 0 defers to the environment");
    setenv("TTIO_THREADS", "junk", 1);
    PASS([TTIOThreads resolve:nil] == 1, "threads: junk resolves to 1");
    unsetenv("TTIO_THREADS");

    int before = ttio_m94z_get_autotune_threads();
    TTIOThreadPool *p1 = [TTIOThreadPool poolWithThreads:1];
    PASS(p1.queue == nil && ttio_m94z_get_autotune_threads() == before,
         "threads: a one-thread pool has no queue and leaves the auto-tune alone");
    [p1 close];
    TTIOThreadPool *p4 = [TTIOThreadPool poolWithThreads:4];
    PASS(p4.queue != nil && p4.queue.maxConcurrentOperationCount == 4,
         "threads: a pool has a queue of its size");
    PASS(ttio_m94z_get_autotune_threads() == 1, "threads: the auto-tune stands down while a pool exists");
    TTIOThreadPool *p2 = [TTIOThreadPool poolWithThreads:2];
    PASS(ttio_m94z_get_autotune_threads() == 1, "threads: nested pools keep it down");
    [p2 close];
    PASS(ttio_m94z_get_autotune_threads() == 1, "threads: still down while the outer pool exists");
    [p4 close];
    PASS(ttio_m94z_get_autotune_threads() == before, "threads: restored at close");

    /* clamp(cores / workers, 2, 8). The floor matters most: one segment
     * thread per block is V6 coding a whole block in sequence. */
    NSUInteger c = (NSUInteger)[[NSProcessInfo processInfo] activeProcessorCount];
    NSUInteger one = [TTIOThreads resolveV6SegmentThreads:1];
    PASS(one == (c > 8 ? 8 : (c < 2 ? 2 : c)),
         "v6 segments: one worker takes the machine, capped at eight");
    PASS([TTIOThreads resolveV6SegmentThreads:c * 2] == 2,
         "v6 segments: more workers than cores still leaves the floor of two");
    PASS([TTIOThreads resolveV6SegmentThreads:0] == one,
         "v6 segments: zero workers is read as one");

    /* The override exists so the split between blocks and segments can
     * be swept. It sits outside the clamp on purpose: a sweep has to be
     * able to ask for 1 and for more than 8. */
    setenv("TTIO_V6_SEGMENT_THREADS", "5", 1);
    PASS([TTIOThreads resolveV6SegmentThreads:1] == 5
         && [TTIOThreads resolveV6SegmentThreads:c * 2] == 5,
         "v6 segments: TTIO_V6_SEGMENT_THREADS wins whatever the worker count");
    setenv("TTIO_V6_SEGMENT_THREADS", "1", 1);
    PASS([TTIOThreads resolveV6SegmentThreads:1] == 1,
         "v6 segments: the override reaches below the floor");
    setenv("TTIO_V6_SEGMENT_THREADS", "16", 1);
    PASS([TTIOThreads resolveV6SegmentThreads:1] == 16,
         "v6 segments: the override reaches above the cap");
    setenv("TTIO_V6_SEGMENT_THREADS", "junk", 1);
    PASS([TTIOThreads resolveV6SegmentThreads:1] == one,
         "v6 segments: junk falls through to the rule");
    setenv("TTIO_V6_SEGMENT_THREADS", "0", 1);
    PASS([TTIOThreads resolveV6SegmentThreads:1] == one,
         "v6 segments: zero falls through to the rule");
    unsetenv("TTIO_V6_SEGMENT_THREADS");
    PASS([TTIOThreads resolveV6SegmentThreads:1] == one,
         "v6 segments: unset falls through to the rule");

    /* The defect this guards: V6 read the auto-tune knob, a pool set
     * that to 1, and V6 has no candidates to race, so every block coded
     * its segments one after another under every writer. */
    int beforeV6 = ttio_m94z_get_v6_threads();
    TTIOThreadPool *p8 = [TTIOThreadPool poolWithThreads:8];
    int inPool = ttio_m94z_get_v6_threads();
    /* Not an equality against resolveV6SegmentThreads:8 — this runs in
     * a suite where an outer pool may already be open, and like the
     * auto-tune the outermost pool is the one that sets the value. What
     * must hold whoever set it is that V6 is not left coding a block's
     * segments one after another, which is the defect this replaced. */
    PASS(inPool > 1,
         "v6 segments: a pool never leaves V6 coding its segments in sequence");
    PASS(inPool >= 2 && inPool <= 8,
         "v6 segments: a pool sets a value inside the clamp");
    TTIOThreadPool *pNested = [TTIOThreadPool poolWithThreads:16];
    PASS(ttio_m94z_get_v6_threads() == inPool,
         "v6 segments: a nested pool leaves the outer count alone");
    [pNested close];
    PASS(ttio_m94z_get_v6_threads() == inPool,
         "v6 segments: closing the nested pool restores nothing");
    [p8 close];
    PASS(ttio_m94z_get_v6_threads() == beforeV6,
         "v6 segments: restored when the pool that set them closes");
}
